ARROW-480: [Python] Implement RowGroupMetaData.ColumnChunk

Licht-T opened the pull request:
This is the patch for [ARROW-480](https://issues.apache.org/jira/projects/ARROW/issues/ARROW-480).

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
+    property has_distinct_count:
+
+        def __get__(self):
+            return bool(self.statistics.has_distinct_count)
```
I didn't think these `bool` casts are needed, but correct me if I'm wrong

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
+        self.statistics.get().Reset()
+
+    def SetComparator(self):
+        self.statistics.get().SetComparator()
```
Are either the Reset method or SetComparator method useful in Python? If not, let's remove them

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
+    def Encode(self):
+        statistics = EncodedStatistics()
+        statistics.init(self.statistics.get().Encode())
+        return statistics
```
Can you use snake case for these functions? `encode_min` etc. If these functions are not useful there's no immediate need to expose them

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
+    property file_path:
+
+        def __get__(self):
+            return self.metadata.file_path()
```
Use `frombytes` here to cast to platform string (for Py3)

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -35,6 +35,271 @@ from pyarrow.lib import ArrowException, NativeFile
+    property path_in_schema:
+
+        def __get__(self):
+            return self.metadata.path_in_schema().get().ToDotString()
```
`frombytes` (though I guess there's no guarantee about UTF8-ness)

wesm commented on python/pyarrow/_parquet.pyx:
```diff
@@ -52,6 +317,11 @@ cdef class RowGroupMetaData:
         self.metadata = self.up_metadata.get()
         self.parent = parent
 
+    def ColumnChunk(self, int i):
```
Maybe just `column`?

wesm commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -442,6 +442,7 @@ def test_parquet_metadata_api():
     # Row group
     rg_meta = meta.row_group(0)
     repr(rg_meta)
+    repr(rg_meta.ColumnChunk(0))
```
Can you add some tests to show that you can obtain the expected column statistics from each column in a row group? Checking the min and max for a string (bytes) column, integer, and double should be good enough. The idea of ARROW-480 was to be able to know when you can skip reading row groups based on a comparison predicate (like a value range)

Licht-T commented:
@wesm Thanks!
- Removed unnecessary class and methods.
- Using Cython `bint` as C++ boolean type.
    http://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#types
- Renamed camel to snake.
- Converting C++ `std::string` to Python `str` by `frombytes`.
- Changed `EncodeMin`/`EncodeMax` to `min`/`max` decoded by `FormatStatValue`.
- Fixed `RowGroupStatistics.__repr__`

Licht-T commented:
Now, I will make new tests!

wesm commented:
hi @Licht-T, will you have time to finish this PR?

Licht-T commented:
@wesm Yes! I'll do this in the couple of days.

wesm commented:
@Licht-T how's it looking on this? Out of the PRs you have open, any that I or @cpcloud can help with?

Licht-T commented:
@wesm Sorry for my late reply. Finally added the tests.

xhochy commented on python/pyarrow/tests/test_parquet.py:
```diff
@@ -469,14 +469,62 @@ def test_parquet_metadata_api():
+        ([-1.1, 2.2, 2.3, None, 4.4], np.float64, u'-1.1', u'4.4', 1, 4),
+        (
+            [u'', u'b', unichar(1000), None, u'aaa'],
+            str, u' ', unichar(1000).encode('utf-8') + ' ', 1, 4
```
This line fails with Python 3, you probably need to change it to `, unichar(1000).encode('utf-8') + b' '`.

Licht-T commented:
@xhochy Thanks! Fixed!
